Implement cat-like filtering behaviour for encrypt/decrypt

amenonsen opened the pull request:
As discussed on IRC, this includes some minor functionality-preserving cleanups for VaultEditor, and uses this to implement the stdin/stdout encryption/decryption semantics requested by @bcoca and @abadger.

amenonsen committed:
Don't create a VaultLib in each method; do it in __init__ instead

amenonsen committed:
Simplify VaultEditor methods

We don't need to keep creating VaultLibs everywhere, and we don't need
to keep checking for errors because VaultLib does it already.

amenonsen committed:
Fix add_option indentation for consistency before adding another option

amenonsen committed:
Implement cat-like filtering behaviour for encrypt/decrypt

This allows the following invocations:

    # Interactive use, like gpg
    ansible-vault encrypt --output x

    # Non-interactive, for scripting
    echo plaintext|ansible-vault encrypt --output x

    # Separate input and output files
    ansible-vault encrypt input.yml --output output.yml

    # Existing usage (in-place encryption) unchanged
    ansible-vault encrypt inout.yml

…and the analogous cases for ansible-vault decrypt as well.

In all cases, the input and output files can be '-' to read from stdin
or write to stdout. This permits sensitive data to be encrypted and
decrypted without ever hitting disk.

abadger commented:
future cleanup note:  VaultLib.cipher_name is not a good attribute.  It's being used in two ways, I think:
- as  a property of the specific file/data.  Since we are able to give different filenames to VaultLib.encrypt()/decrypt() now, having a VaultLib level property specifying the cipher of the last data that went through VaultLib doesn't make sense.  Probably better to replace this with a method that returns the cipher_name when given a ciphertext.
- to tell us what cipher to use to encrypt a file.  Because this belongs to the specific file to be encrypted and that can change for each call to encrypt()/decrypt() we probably should pass this in as a parameter to encrypt() instead.

I can't think of any way that self.cipher_name will break current code so we don't have to change it in this PR but it seems conceptually wrong and might be a source of bugs with future code changes.

bcoca commented on lib/ansible/cli/vault.py:
```diff
@@ -63,7 +63,19 @@ def parse(self):
+            if len(self.args) > 1:
+                raise AnsibleOptionsError("At most one input file may be used with the --output option")
+
+        elif len(self.args) == 0:
```
so even if i try to input from stdin this will error out?

abadger commented:
Review and everything here looks good to me.  At first I had some questions in the same area as @bcoca did but I think all of the ways of operating that I'd expect are covered.  Just not quite the same way as cat:

```
$ ansible-vault encrypt filename1   # std way we've always called the command
$ ansible-vault encrypt filename1 -o filename2    # New cmdline arg
$ cat filename1 | ansible-vault encrypt -o filename2   # Stdin to named file
$ ansible-vault encrypt filename1 -o - > filename2     # Output to stdout
$ cat filename1 | ansible-vault encrypt filename1 -o - > filename2  # stdin to stdout
```

There's other UNIX tools that operate this way (using explicit -o- to specify output to stdout) so it looks okay to me.  @bcoca does that summary of the options alleviate your concern?

amenonsen committed:
Also support output to stdout with no arguments

This allows "cat plaintext|ansible-vault encrypt > ciphertext".

amenonsen committed:
More helpful prompts from ansible-vault encrypt/decrypt

Now we issue a "Reading … from stdin" prompt if our input isatty(), as
gpg does. We also suppress the "x successful" confirmation message at
the end if we're part of a pipeline.

(The latter requires that we not close sys.stdout in VaultEditor, and
for symmetry we do the same for sys.stdin, though it doesn't matter in
that case.)

amenonsen committed:
Update ansible-vault manpage to describe new encrypt/decrypt behaviour

abadger commented:
Nice work!  I like the new behaviour a lot.  thanks @amenonsen !

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
